packages/esm/core/src/Engines/engine.adapters.ts

```diff
@@ -0,0 +1,46 @@
+/**
+ * Generate adapters from existing engine state object.
+ * Adapters can be used to simulate a real engine object.
+ */
+
+import type { ThinEngine } from "@babylonjs/core/Engines/thinEngine.js";
+import type * as baseTypes from "./engine.base.js";
+import type * as webGLTypes from "./WebGL/engine.webgl.js";
+import type * as webGPUTypes from "./WebGPU/engine.webgpu.js";
+
+type PickMatching<T, V> = { [K in keyof T as T[K] extends V ? K : never]: T[K] };
+type ExtractMethods<T> = PickMatching<T, Function>;
+
+type EngineMethods<T extends ThinEngine> = ExtractMethods<T>;
+type BaseThinEngineUnionMethods<T extends ThinEngine, E> = PickMatching<EngineMethods<T>, E>;
+
+export type BaseEngineMethods = ExtractMethods<typeof baseTypes>;
+export type WebGLEngineMethods = ExtractMethods<typeof webGLTypes>;
+export type WebGPUEngineMethods = ExtractMethods<typeof webGPUTypes>;
+
+/**
+ * Augment an engineState object with methods to simulate a real engine object
+ *
+ * @param engineState the engineState object to augment
+ * @param injectedMethods The methods that will be injected to the engineState
+ * @param force Should we force re-injecting the methods
+ * @returns The engineState cased to the requested engine type
+ */
+export function augmentEngineState<T extends ThinEngine, E = BaseEngineMethods | WebGLEngineMethods | WebGPUEngineMethods>(

```

Comment:
Why do you need to inject methods in the engine state when (it seems!) you don't use them? When a method of the engine must be called, you are doing `method_engine(ps, ...)` with ps=engine state, and not `ps.method_engine(...)`? But maybe I missed something, I didn't review everything.

Reply:
it is true, i don't use them in esm, but this way the engine state is turning to a full engine and can be used in other components that require an engine. This way effect, internal textures and so on can still work without changing the types they accept in the construtor
I have considered different ways to implement the "legacy"  support. performance-wise, this was the best way, as it doesn't involve accessing extra objects or arrays.